### PR TITLE
Yehor/pred 1342 fix dialect detection

### PR DESCRIPTION
**Pull request by nimnull**
Brief intro:
Added own parser to detect features delimiter.
Made resampling to feed `csv.sniffer` only with completed lines (last line could be truncated by default)

### CODE REVIEW

File: datarobot_batch_scoring/detect.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        resampled = "".join(lines_content[:-1])
+        return self.analyze(lines_analyzed), resampled
+
+    def get_sample(self, sample, sample_lines=20, quotechar='"'):
+        enclosed = False
+        actual_lines = 1
+        lines = []
+        single_line = ""
+
+        # raise Exception("Sample size %d" % len(sample))
```
Comment (Axik) [inline datarobot_batch_scoring/detect.py]: unclear what this comment line stands for

File: datarobot_batch_scoring/detect.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        for idx, ch in enumerate(sample):
+            if actual_lines >= sample_lines:
+                break
+            prev = sample[idx-1] if idx else None
+            next = sample[idx+1] if (idx + 1) < len(sample) else None
```
Comment (Axik) [inline datarobot_batch_scoring/detect.py]: built-in function re-assignment 👎 ,
especially when we're using this built-in in other part of application, like reader.

File: datarobot_batch_scoring/detect.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import math
+import re
+
+
+class Detector(object):
```
Comment (Axik) [inline datarobot_batch_scoring/detect.py]: small doc string about what this class supposed to do would be awesome.
It is not very clear does it detect delimiters or encoding, for example

File: datarobot_batch_scoring/detect.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                single_line = ""
+
+            elif not enclosed and not self.non_delimiter_re.match(ch):
+                self.increment(ch, actual_lines)
+
+        return actual_lines, lines
+
+    def analyze(self, lines_analyzed):
+        candidates = []
+        for delim, freq in list(self.frequency_table.items()):
```
Comment (Axik) [inline datarobot_batch_scoring/detect.py]: I thought items() returning iterable in both cases, and we don't need to list evaluate it.

**Comment by romankolpak** [comment]
I'm a bit concerned about introducing our own heuristic to detect dialects to be honest. Automatic detection of dialects is a hard problem, especially for larger datasets. 

Maybe we could provide a more fine-grained control over the dialect properties to the user instead? An example of that would be `pandas.read_csv` which provides an ability to override stuff like `qoutechar` and `lineterminator` via its arguments. I feel like that's a more robust solution, as we are unlikely to cover all the edge-cases with home-grown detector. 

Let me know what you think.

File: datarobot_batch_scoring/detect.py (hunk 1 above)
Comment (nimnull) [inline datarobot_batch_scoring/detect.py]: debug, did not dropped it yet

**Comment by nimnull** [comment]
@romankolpak 
If you read this code clearly, I just introduced resampling to feed csv.Sniffer with complete set of lines. Without truncated one. Also. My detector works only as a doom's day machine, when automatic detection failed. 
Both approaches use only a limited size sample for detection. So this is no difference between 10Mb and 1T files. 
Pandas is prohibited due to implicit column type conversions. We did parse with pandas many versions earlier, but abandoned that approach because of that 'types' reason.

File: datarobot_batch_scoring/detect.py (hunk 3 above)
Comment (nimnull) [inline datarobot_batch_scoring/detect.py]: it is obvious, cause there is no calls to `chardet` and it used at the point, when we already know data encoding (:. 
Just delimiters. Also, it's main method return value called `delimiter`.

File: datarobot_batch_scoring/detect.py (hunk 4 above)
Comment (nimnull) [inline datarobot_batch_scoring/detect.py]: yep, nice thought. Missed to clean that part after refactored resampling.

File: datarobot_batch_scoring/detect.py (hunk 3 above)
Comment (Axik) [inline datarobot_batch_scoring/detect.py]: I mean, you have to thoroughly read the code, comments/docs string made for common-sense obvious information that gives heads up before code readeing.

**Pull request merged**